### DISCUSSION TITLE
Questions about syntax highlighting creation

### DISCUSSION THREAD

**Discussion by katawful** [tree-sitter/tree-sitter#1142] [Q&A] [answered]
I'm finally at the point in my tree-sitter parser where I can safely create the syntax highlighter, but the documentation for it it leaves me confused. I have multiple questions because of this:

1. Why isn't config.json in repos? The docs say its needed to use the highlighting, but none of the parser repos have it
2. When I run `tree-sitter highlight file.extension` it tells me it couldn't find the language for the file I pass. Am I missing something? I added the appropriate file type to package.json
3. Can I import files into my highlight.scm? There are a lot of keywords I would like to catch and I don't want to crowd my main highlight file

**Comment by ahlinc** [comment, marked as answer]
By your questions:
1. The `config.json` file it's the global file that exists in the `$HOME/.tree-sitter/` folder. If it's correctly configured than you can use `tree-sitter highlight` command from any place.
2. Perhaps you missed to configure a path to your language folder.
3. No, this is not supported, it may contain only s-exp queries and comments between queries that start with the `;` symbol.

To do highlighting you need to:
1. Run `tree-sitter generate` that will generate the parser.
2. Run [`tree-sitter init-config`](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#global-configuration) and it will generate a config file in `$HOME/.tree-sitter/config.json`, that's your `config.json` file template. **Be careful:** if you've made changes to these file before, it'll will be silently overwritten to the initial state.
3. Add path to the base folder that contains your `tree-sitter-LANGUAGE` grammar folder in the `config.json` file by inserting it into the `"parser-directories"` list.
4. Verify that tree-sitter can see your parser by the `tree-sitter dump-languages` command.
5. Ensure that there is `queries/highlights.scm` file in your `tree-sitter-LANGUAGE` grammar folder.
6. Configure matching for source files related to your language in the [`package.json`](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#example) file that exist in your grammar folder.
7. Try to parse something with command like `tree-sitter highlight source_file.EXT` where `EXT` is the one of file extensions configured on previous step.
8. Adjust `"theme"` dictionary in the `config.json` file to not-terminals from your grammar if you see no colors or wrong / not enough colors on previous step.

**Reply by katawful** [reply]
I can't seem to get my languages to be detected despite having them in my config. When I run `tree-sitter dump-languages` from my language working folder it doesn't return anything

**Reply by ahlinc** [reply]
Could you provide more info about your environment?
* What OS do you use?
* Content of the `config.json` file and your local full path to it.
* Your local full path to the `tree-sitter-obse` folder.

---

When I cloned your `tree-sitter-obse` repo to my grammars folder, the output from the`tree-sitter dump-languages` command has this:
```
scope:
parser: "/home/user1/repos/tree-sitter/grammars/tree-sitter-obse"
highlights: None
file_types: []
content_regex: None
injection_regex: None
```
After adding more config from the below patch it has this:
```
scope: source.obse
parser: "/home/user1/repos/tree-sitter/grammars/tree-sitter-obse/"
highlights: None
file_types: ["obse"]
content_regex: None
injection_regex: None
```

`# cat ~/.tree-sitter/config.json`
```json
{
  "parser-directories": [
    "/home/user1/repos/tree-sitter/grammars"
  ],
  "theme": {
    "variable.parameter": {
      "underline": true
    },
    "type.builtin": {
      "bold": true,
      "color": 23
    },
    "attribute": {
      "italic": true,
      "color": 124
    },
    "number": {
      "bold": true,
      "color": 94
    },
    "comment": {
      "color": 245,
      "italic": true
    },
    "function": 26,
    "operator": {
      "bold": true,
      "color": 239
    },
    "function.builtin": {
      "bold": true,
      "color": 26
    },
    "string": 28,
    "punctuation.bracket": 239,
    "keyword": 56,
    "punctuation.delimiter": 239,
    "string.special": 30,
    "tag": 18,
    "type": 23,
    "constant": 94,
    "variable.builtin": {
      "bold": true
    },
    "property": 124,
    "embedded": null,
    "constructor": 136,
    "constant.builtin": {
      "color": 94,
      "bold": true
    }
  }
}
```





**Patch for the repo:**
```diff
diff --git a/package.json b/package.json
index eb83c8b..403a562 100644
--- a/package.json
+++ b/package.json
@@ -25,5 +25,13 @@
   },
   "devDependencies": {
     "tree-sitter-cli": "^0.19.4"
-  }
+  },
+  "tree-sitter": [
+    {
+      "scope": "source.obse",
+      "file-types": [
+        "obse"
+      ]
+    }
+  ]
 }
diff --git a/queries/highlights.scm b/queries/highlights.scm
new file mode 100644
index 0000000..d832d1c
--- /dev/null
+++ b/queries/highlights.scm
@@ -0,0 +1,12 @@
+[
+  "begin"
+  "foreach"
+  "loop"
+  "scn"
+  "end"
+  "eval"
+]
+@keyword
+
+container: (identifier) @attribute
+
```

### Highlighting sample:
![image](https://user-images.githubusercontent.com/14666676/120906638-ca1ea500-c663-11eb-966c-03905e91c5e2.png)

**Reply by ahlinc** [reply]
By the way, if you want to distinguish numbers `2:28` from strings `5:28` you need to add named rules for those and use thus rules in the `literal` rule, this will allow you to address those items in queries.

**Reply by katawful** [reply]
I was dumb and misread the info about the parser directory. I added the directory for each filetype, not the folder containing them. It does what I want now, but when I go to highlight a test file like you came up with it doesn't return anything and just locks in execution like its waiting. Is it supposed to take some amount of time?

I'm running on Arch Linux and used npm to install tree-sitter if that helps

**Reply by ahlinc** [reply]
Yes, it takes some time for the first run cause it generates binary grammar file in the `~/.tree-sitter/bin/obse.so` path.

**Reply by katawful** [reply]
The file is created but the highlight command still doesn't respond

**Reply by ahlinc** [reply]
Hmm, what happens on CPU, does it consumes one core to 100% or not?
Could you tell your `tree-sitter-cli` version and try to build `tree-sitter-cli` from the master branch of this repo and test with it?

**Reply by katawful** [reply]
Must have been a bug in the release I was using. Updating with npm solved it. Thank you for the help!

One last thing, is there a list of general highlighting types? Like `@keyword`, `@type`, `@string`, etc...

**Reply by ahlinc** [reply]
I don't know about the general list, I saw thoughts about standardization of such list but nothing more.
You can use existing keywords in the `config.json` and extend that list for your own needs.
Also for impression you may look to nvim's queries: https://github.com/nvim-treesitter/nvim-treesitter/tree/master/queries
Or look to the base list of keywords from vscode: https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#predefined-textmate-scope-mappings

**Reply by katawful** [reply]
Alright, thank you!
